### ISSUE TITLE
Azure cloud provider config file

### ISSUE THREAD

**Issue by DmitriyProkhorov** [apprenda/kismatic#1183]
**Is this a BUG REPORT or FEATURE REQUEST?**: HowTo or bug

**What happened**:
I want to install a kubernetes cluster in Azure. I use KET for deploy to Azure virtual machines (I do not want to use AKS Azure, which is now in the preview)
Everything is installed and works well until I turn on the *cloud_provider: "azure"* option in the kismatic-cloud.yml file. There is little information on the Internet about what a configuration file should be for an azure provider option. I found only the one documentation for the link https://github.com/alexbrand/ket-on-azure, where there is an example of a configuration, but this does not work for KET. I get an error:

**What you expected to happen**:

Validate Kubernetes Control Plane is Running                                    [ERROR]
- Task: fail if pod 'kube-apiserver' does not have the desired version
  k8s-master1-lab: An error occurred getting the pod's version using kubectl.

The connection to the server 127.0.0.1:6443 was refused - did you specify the right host or port?
error executing jsonpath "{.items[0].metadata.annotations.kismatic/version}": array index out of bounds: index 0, length 0
                                                                                [ERROR]
  k8s-master2-lab: An error occurred getting the pod's version using kubectl.

The connection to the server 127.0.0.1:6443 was refused - did you specify the right host or port?
error executing jsonpath "{.items[0].metadata.annotations.kismatic/version}": array index out of bounds: index 0, length 0
                                                                                [ERROR]
error installing: error running playbook: error running ansible: exit status 2

I need help with the correct instruction that you need to specify in the file azure_cloud_provider.config
**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

What I did:
1. Deployed Azure infrastructure: 3 etcd, 2 master, 2 worker, 2 ingress
- all in one resource group
2. The file azure_cloud_provider.config:
{
    "tenantId": "e70xxx-xxxx-xxxx-xxxx-xxxx4774",
    "subscriptionId": "1524xxx-xxxx-xxxx-xxxx-xxxxx2fa",
    "aadClientId": "5xxxx-xxxx-xxxx-xxxx-xxxxx991bcb",
    "aadClientSecret": "xxxxx-xxxxx-xxxx-xxxx-xxxxxa5",
    "resourceGroup": "RG-k8s-lab"
}

**Environment**:
- Kismatic version (use `kismatic version`): 1.9.1
- Plan file Gist (remove sensitive information if any): 
in the attachment
[kismatic-cluster.zip](https://github.com/apprenda/kismatic/files/1865791/kismatic-cluster.zip)
- Cloud provider or hardware configuration: Azure
- OS (e.g. from /etc/os-release): CentOS 7.3
- Kernel (e.g. `uname -a`): Linux k8s-master1-lab 3.10.0-514.26.2.el7.x86_64 #1 SMP Tue Jul 4 15:04:05 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

**Comment by emmetthitz** [comment]
Which CNI are you using?

**Comment by dkoshkin** [comment]
Check the `kube-apiserver` logs by SSHing on the master and running `docker logs` on the `kube-apiserver` container.

**Comment by DmitriyProkhorov** [comment]
CNI - weave

**Comment by DmitriyProkhorov** [comment]
I checked logs and see the following:
I0330 20:20:46.454028       1 server.go:121] Version: v1.9.3
I0330 20:20:46.454082       1 cloudprovider.go:59] --external-hostname was not specified. Trying to get it from the cloud provider.
I0330 20:20:46.454534       1 azure.go:249] azure: using client_id+client_secret to retrieve access token
E0330 20:20:46.466424       1 azure_blobDiskController.go:76] azureDisk - getAllStorageAccounts error: azure.BearerAuthorizer#WithAuthorization: Failed to refresh the Token for request to https://management.azure.com/subscriptions/1524dbf7-2965-4d78-9654-0866a30352fa/providers/Microsoft.Storage/storageAccounts?api-version=2017-06-01: StatusCode=0 -- Original Error: adal: Failed to execute the refresh request. Error = 'Post https://login.microsoftonline.com/e70cc890-1be6-4d4f-98f8-84204d344774/oauth2/token?api-version=1.0: x509: failed to load system roots and no roots provided'
E0330 20:20:46.486109       1 azure_util.go:421] error: az.getIPForMachine(k8s-master1-lab), az.getVirtualMachine(k8s-master1-lab), err=azure.BearerAuthorizer#WithAuthorization: Failed to refresh the Token for request to https://management.azure.com/subscriptions/1524dbf7-2965-4d78-9654-0866a30352fa/resourceGroups/RG-k8s-lab/providers/Microsoft.Compute/virtualMachines/k8s-master1-lab?%24expand=instanceView&api-version=2017-03-30: StatusCode=0 -- Original Error: adal: Failed to execute the refresh request. Error = 'Post https://login.microsoftonline.com/e70cc890-1be6-4d4f-98f8-84204d344774/oauth2/token?api-version=1.0: x509: failed to load system roots and no roots provided'
E0330 20:20:46.486128       1 azure_backoff.go:141] backoff: failure, will retry,err=azure.BearerAuthorizer#WithAuthorization: Failed to refresh the Token for request to https://management.azure.com/subscriptions/1524dbf7-2965-4d78-9654-0866a30352fa/resourceGroups/RG-k8s-lab/providers/Microsoft.Compute/virtualMachines/k8s-master1-lab?%24expand=instanceView&api-version=2017-03-30: StatusCode=0 -- Original Error: adal: Failed to execute the refresh request. Error = 'Post https://login.microsoftonline.com/e70cc890-1be6-4d4f-98f8-84204d344774/oauth2/token?api-version=1.0: x509: failed to load system roots and no roots provided'
I0330 20:20:46.486137       1 azure_instances.go:53] NodeAddresses(k8s-master1-lab) abort backoff
error setting the external host value: failed to get external host address from "azure" cloud provider: timed out waiting for the condition

**Comment by dkoshkin** [comment]
What OS are you using?

**Comment by DmitriyProkhorov** [comment]
CentOS 7.3. I updated the environment settings in the issue ticket. Sorry that I did not do it at first

**Comment by DmitriyProkhorov** [comment]
I'm a little advanced in my problem! I deployed the original cluster AKS from Microsoft, looked at the configuration file (azure_cloud_provider.config) there and made it for KET ( in AKS file name is azure.json). The error has disappeared. The cluster started. But there is another problem. For example, when I want to deploy a persistant volume, I get an error:

C:\Users\prokh>kubectl describe pvc
Name:          volume-claim-name
Namespace:     default
StorageClass:  premium
Status:        Pending
Volume:
Labels:        <none>
Annotations:   kubectl.kubernetes.io/last-applied-configuration={"apiVersion":"v1","kind":"PersistentVolumeClaim","metadata":{"annotations":{"volume.beta.kubernetes.io/storage-class":"premium"},"name":"volume-claim-...
               volume.beta.kubernetes.io/storage-class=premium
               volume.beta.kubernetes.io/storage-provisioner=kubernetes.io/azure-disk
Finalizers:    []
Capacity:
Access Modes:
Events:
  Type     Reason              Age   From                         Message
  ----     ------              ----  ----                         -------
  Warning  ProvisioningFailed  2m    persistentvolume-controller  Failed to provision volume with StorageClass "premium": azure.BearerAuthorizer#WithAuthorization: Failed to refresh the Token for request to https://management.azure.com/subscriptions/1xxxxx-xxxxx-xxxx-xxx-08xxxxxxxxx/resourceGroups/RG-k8s-lab/providers/Microsoft.Compute/disks/kubernetes-dynamic-pvc-xxxxxxeb-xxx-xxxx-xxx-xxxxxxxxx?api-version=2016-04-30-preview: StatusCode=0 -- Original Error: adal: Failed to execute the refresh request. Error = 'Post https://login.microsoftonline.com/xxxxx-xxxx-xxxx-xxxx-xxxxx/oauth2/token?api-version=1.0: x509: failed to load system roots and no roots provided'

For some reason, azure authentication to azur api does not work

**Comment by DmitriyProkhorov** [comment]
All the same the cloud provider for azure does not work. Re-created the environment again. Everything from scratch. The error repeated:

E0402 09:37:33.504889       1 azure_backoff.go:141] backoff: failure, will retry,err=azure.BearerAuthorizer#WithAuthorization: Failed to refresh the Token for request to https://management.azure.com/subscriptions/xxxx-xxxx-xxx-xxxx-xxxxx/resourceGroups/RG-k8s-lab/providers/Microsoft.Compute/virtualMachines/k8s-master1-lab?%24expand=instanceView&api-version=2017-03-30: StatusCode=0 -- Original Error: adal: Failed to execute the refresh request. Error = 'Post https://login.microsoftonline.com/xxx-xxxx-xxx-xxx-xxxxxxx/oauth2/token?api-version=1.0: x509: failed to load system roots and no roots provided'
I0402 09:37:33.504899       1 azure_instances.go:35] NodeAddresses(k8s-master1-lab) abort backoff
error setting the external host value: failed to get external host address from "azure" cloud provider: timed out waiting for the condition

Perhaps this is the problem of the Azure cloud provider in kubernetes. Can someone confirm this?

**Comment by dkoshkin** [comment]
@DmitriyProkhorov looks like the OS image you are using does not use standard CA certificates location `/etc/ssl/certs/`, the `kube-apiserver` and `controller-manager` will need those to be able to connect to the Azure's API.

Can you take a look for your specific OS where those certs are and then mount those in the ansible playbooks below to see if that fixes it:
```
ansible/playbooks/roles/kube-apiserver/templates/kube-apiserver.yaml
ansible/playbooks/roles/kube-controller-manager/templates/kube-controller-manager.yaml
```

**Comment by dkoshkin** [comment]
Interesting from the [issue you created upstream](https://github.com/kubernetes/cloud-provider-openstack/issues/74) they are suggesting to remove `/etc/ssl`

**Comment by DmitriyProkhorov** [comment]
On Ubuntu Server it all worked. The way to clean the / etc / ssl folder did not help for me.
Can I use CentOS 7.4? The requirements of KET specify Centos 7.3 only

**Comment by dkoshkin** [comment]
We test and recommend `7.3` because thats the version available on AWS, however `7.4` should work.

**Comment by dkoshkin** [comment]
@DmitriyProkhorov I'm curious to know if 7.4 worked without having to modify any of the playbooks.
